### PR TITLE
Softmax free sampling

### PR DESCRIPTION
**Pull request by kf-zhang**
Using the Gumbel-max trick to accelerate the sampling process. This allows direct sampling from logits, eliminating the need to first convert logits into a probability distribution via the softmax function before sampling.
In torch, the Gumbel-max trick can be implemented as follows:
```python
# Instead of:
probs = torch.softmax(logits, dim=-1)
next_token = torch.multinomial(probs, num_samples=1)

# We can do:
gumbel_noise = -torch.log(-torch.log(torch.rand_like(logits)))
next_token = torch.argmax(logits + gumbel_noise, dim=-1)
```

To verify the correctness of the Gumbel-max trick implementation, run the following test:
```bash
pytest ./tests/test_sampling.py::test_sampling_from_logits
pytest ./tests/test_sampling.py::test_sampling_from_logits_freq
```

To benchmark the performance of the Gumbel-max trick against the traditional softmax sampling method, run:
```bash
python ./benchmarks/bench_sampling.py
```

To see the theoretical background of the Gumbel-max trick, refer to the following link:
[Accelerating LLM Inference: Fast Sampling with Gumbel-Max Trick](https://huggingface.co/blog/cxdu/fastsampling)
[A Review of the Gumbel-max Trick and its Extensions for Discrete Stochasticity in Machine Learning](https://arxiv.org/pdf/2110.01515)
[FLASHSAMPLING: FAST ANDMEMORY-EFFICIENT EXACTSAMPLINGWITHGROUP-GUMBEL-MAX](https://openreview.net/pdf?id=V4Xs283LHH)

### CODE REVIEW

File: include/flashinfer/sampling.cuh [hunk 1]
```diff
@@ -339,6 +340,102 @@ __device__ __forceinline__ void DeviceSamplingFromProb(
...
+                                                                      uint64_t philox_offset,
+                                                                      uint64_t subsequence) {
+  curandStatePhilox4_32_10_t state;
+  vec_t<float, VEC_SIZE> noise;
+// TODO: compare the speed of log2 and log
+#pragma unroll
+  for (uint32_t i = 0; i + 4 <= VEC_SIZE; i += 4) {
+    curand_init(philox_seed, subsequence + i, philox_offset, &state);
+    float4 noise_vec = curand_uniform4(&state);
+    noise[i] = -log(-log(noise_vec.x));
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: log(x) will be lowered to log2(x) * log(2), so it's better to use log2 directly.

Also, should we add some eps term to avoid extreme case? (https://gist.github.com/yzh119/fd2146d2aeb329d067568a493b20172f#file-st-gumbel-py-L9)

Comment (kf-zhang) [reply include/flashinfer/sampling.cuh]: Great suggestion. The modifications have been made in the new code.

File: include/flashinfer/sampling.cuh [hunk 2]
```diff
@@ -339,6 +340,116 @@ __device__ __forceinline__ void DeviceSamplingFromProb(
...
+
+template <typename DType, uint32_t VEC_SIZE>
+__device__ __forceinline__ vec_t<DType, VEC_SIZE> GenerateGumbelNoise(uint64_t philox_seed,
+                                                                      uint64_t philox_offset,
+                                                                      uint64_t subsequence) {
+  curandStatePhilox4_32_10_t state;
+  vec_t<float, VEC_SIZE> noise;
+  constexpr float kEPSILON = 1e-20f;
+  constexpr float kLOG2 = 0.6931471806f;
+  auto log_func = [](float x) { return log2f(x) * kLOG2; };
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: You don't have to do this explicitly, compiler will do it automatically.

What do I mean by using log2 is that:
Suppose c = log2(e)
-log(-log(x)) = -log2(-log2(x) * c) * c = (-log2(-log2(x)) + -log2(c)) * c = -log2(-log2(x)) * c - (c * log2(c))

The last term c * log2(c) is a constant term that you can safely ignore, so you can just use -log2(-log2(x)) * c.

After adding epsilon, it become  -log2(-log2(x + eps) + eps) * c

File: csrc/sampling.cu [hunk 3]
```diff
@@ -25,6 +25,33 @@
...
+  auto device = logits.device();
+  CHECK_DIM(2, logits);  // logits: (batch_size, vocab_size)
+  unsigned int batch_size = output.size(0);
+  unsigned int vocab_size = logits.size(1);
+
+  uint64_t philox_seed, philox_offset;
+  auto gen = at::get_generator_or_default<at::CUDAGeneratorImpl>(
+      gen_, at::cuda::detail::getDefaultCUDAGenerator());
+  std::lock_guard<std::mutex> lock(gen->mutex_);
+  at::PhiloxCudaState rng_engine_inputs = gen->philox_cuda_state(batch_size);
```
Comment (yzh119) [inline csrc/sampling.cu]: Set `increment` to `batch_size * vocab_size` because we have to generate large number of uniform numbers.
```suggestion
  at::PhiloxCudaState rng_engine_inputs = gen->philox_cuda_state(batch_size * vocab_size);
```

See https://github.com/pytorch/pytorch/blob/b34146a093ed868d0cb7de63c094f847cf82ad0b/aten/src/ATen/cuda/CUDAGeneratorImpl.cpp#L432-L442 for the explaination of `increment`.
